### PR TITLE
ROX-31454: adds node file event detection

### PR DESCRIPTION
**Pull request by Stringy**
## Description

Adds the full node file activity detection, including plumbing into Sensor and the file system pipeline.

It currently only supports Node events but where appropriate, logic for deployment based events has been added (i.e. in functions that should operate on generic file activity events/alerts and not node-specific ones)

After this PR is merged, we have full end-to-end detection and alerting for Node file system events. In future PRs we will build on the existing testing to provide e2e testing of this feature.

## User-facing documentation

- [ ] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [ ] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [x] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

### How I validated my change

I have tried to implement thorough unit tests for the detection of file system events for a given node. These test cases have been created with some assistance from Claude, and have been verified for correctness and usefulness. The assistance was less about code generation and more about identifying key test cases and corner cases which are worth testing for.

### CODE REVIEW

File: pkg/detection/runtime/detector.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ type Detector interface {
 	DetectForDeploymentAndKubeEvent(enhancedDeployment booleanpolicy.EnhancedDeployment, kubeEvent *storage.KubernetesEvent) ([]*storage.Alert, error)
 	DetectForDeploymentAndNetworkFlow(enhancedDeployment booleanpolicy.EnhancedDeployment, flow *augmentedobjs.NetworkFlowDetails) ([]*storage.Alert, error)
 	DetectForAuditEvents(auditEvents []*storage.KubernetesEvent) ([]*storage.Alert, error)
+	DetectForNodeAndFileAccess(node *storage.Node, access *storage.FileAccess) ([]*storage.Alert, error)
```
Comment (clickboo) [inline pkg/detection/runtime/detector.go:20]: Nit: DetectForNodeFileAccess

File: pkg/detection/runtime/detector.go [hunk 2]
```diff
@@ -69,6 +70,41 @@ func (d *detectorImpl) DetectForDeploymentAndNetworkFlow(
...
+		if compiled.Policy().GetDisabled() {
+			return nil
+		}
+
+		if access != nil {
+			// Check predicate on file access.
+			if !compiled.AppliesTo(access) {
+				return nil
+			}
+
```
Comment (clickboo) [inline pkg/detection/runtime/detector.go:87]: Would be nice to add a TODO here to add a check for scopes/exclusions here when we support them for node policies

File: pkg/detection/runtime/utils.go [hunk 3]
```diff
@@ -25,6 +25,31 @@ func constructProcessAlert(policy *storage.Policy, deployment *storage.Deploymen
 	return alert
 }
 
+func constructFileAccessAlert(policy *storage.Policy, node *storage.Node, deployment *storage.Deployment, violations booleanpolicy.Violations) *storage.Alert {
```
Comment (clickboo) [inline pkg/detection/runtime/utils.go:31]: This takes node and deployment both because it builds a file access alert for both paths - assuming one will be nil, in each of the two paths.

File: pkg/detection/runtime/utils.go [hunk 4]
```diff
@@ -25,6 +25,31 @@ func constructProcessAlert(policy *storage.Policy, deployment *storage.Deploymen
...
+	if len(violations.AlertViolations) == 0 && violations.FileAccessViolation == nil {
+		return nil
+	}
+
```
Comment (clickboo) [inline pkg/detection/runtime/utils.go:35]: a comment like the one I have above would be nice before this check

File: pkg/detection/runtime/utils.go [hunk 5]
```diff
@@ -25,6 +25,31 @@ func constructProcessAlert(policy *storage.Policy, deployment *storage.Deploymen
...
+	} else {
+		alert = constructGenericRuntimeAlert(policy, deployment, violations.AlertViolations)
+		if action, msg := buildEnforcement(policy); action != storage.EnforcementAction_UNSET_ENFORCEMENT {
+			alert.Enforcement = &storage.Alert_Enforcement{
+				Action:  action,
+				Message: msg,
+			}
+		}
+	}
+	alert.FileAccessViolation = violations.FileAccessViolation
```
Comment (clickboo) [inline pkg/detection/runtime/utils.go:53]: note to self: what is this field?

File: pkg/detection/compiled_policy.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ type CompiledPolicy interface {
 	MatchAgainstKubeResourceAndEvent(cacheReceptacle *booleanpolicy.CacheReceptacle, kubeEvent *storage.KubernetesEvent, kubeResource interface{}) (booleanpolicy.Violations, error)
 	MatchAgainstAuditLogEvent(cacheReceptacle *booleanpolicy.CacheReceptacle, kubeEvent *storage.KubernetesEvent) (booleanpolicy.Violations, error)
 	MatchAgainstDeploymentAndNetworkFlow(cacheReceptable *booleanpolicy.CacheReceptacle, enhancedDeployment booleanpolicy.EnhancedDeployment, flow *augmentedobjs.NetworkFlowDetails) (booleanpolicy.Violations, error)
+	MatchAgainstNodeAndFileAccess(cacheReceptacle *booleanpolicy.CacheReceptacle, node *storage.Node, access *storage.FileAccess) (booleanpolicy.Violations, error)
```
Comment (clickboo) [inline pkg/detection/compiled_policy.go:27]: Nit: MatchAgainstNodeFileAccess

File: pkg/detection/compiled_policy.go [hunk 7]
```diff
@@ -62,10 +63,16 @@ func newCompiledPolicy(policy *storage.Policy) (CompiledPolicy, error) {
 				"one runtime constraint from either process, or kubernetes event category, or network baseline.", policy.GetName())
 		}
 		// set predicates
-		compiled.predicates = append(compiled.predicates, &deploymentPredicate{scopes: scopes, exclusions: exclusions})
+		// Only add deployment predicate for deployment-based runtime events, not for node events
+		if policy.GetEventSource() != storage.EventSource_NODE_EVENT {
```
Comment (clickboo) [inline pkg/detection/compiled_policy.go]: This might be more readable and easier to understand as a switch case with the order of cases reversed as compared to the ifs.
case storage.EventSource_NODE_EVENT:
case storage.EventSource_AUDIT_LOG_EVENT:
default

File: pkg/detection/compiled_policy.go
Comment (clickboo) [inline pkg/detection/compiled_policy.go]: same for this: switch case instead of multiple ifs.

File: sensor/common/detector/queue/file_accesses.go [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package queue
+
+import (
+	"context"
+
+	"github.com/stackrox/rox/generated/storage"
+)
+
+type FileAccessQueueItem struct {
+	Ctx        context.Context
```
Comment (clickboo) [inline sensor/common/detector/queue/file_accesses.go:10]: It does worry me that we use the same storage proto in the communication/internal API protos. Given our backward compatibility requirement with sensor versions supported by central version, I would recommend copies in the internal API proto layer.

File: sensor/common/detector/detector.go [hunk 9]
```diff
@@ -101,6 +103,14 @@ func New(clusterID clusterIDPeekWaiter, enforcer enforcer.Enforcer, admCtrlSetti
 		detectorMetrics.DetectorDeploymentDroppedCount,
 	)
 
+	fileAccessQueue := queue.NewQueue[*queue.FileAccessQueueItem](
+		detectorStopper,
+		"FilesQueue",
```
Comment (clickboo) [inline sensor/common/detector/detector.go]: FileAccessQueue

File: sensor/common/detector/detector.go [hunk 10]
```diff
@@ -822,3 +838,75 @@ func (d *detectorImpl) processNetworkFlow(ctx context.Context, flow *storage.Net
...
+		case item, ok := <-d.fileAccessQueue.Pull():
+			if !ok {
+				return
+			}
+			if item == nil {
+				continue
+			}
+
+			var alerts []*storage.Alert
+			if item.Node != nil {
```
Comment (clickboo) [inline sensor/common/detector/detector.go]: It might improve code readability a lot if we have some helper methods like isNodeFileAccess and isDeploymentFileAccess on the actual file access event proto - instead of using nil checks

File: pkg/detection/runtime/utils.go (hunk 5 above)
Comment (Stringy) [reply pkg/detection/runtime/utils.go:53]: This is a new field, modeled on ProcessViolation, which contains one or more file accesses so that we can merge alerts for the same entity while keeping all the accesses performed by that entity

File: pkg/detection/compiled_policy.go (hunk 6 above)
Comment (Stringy) [reply pkg/detection/compiled_policy.go:27]: Yeah I think I'd prefer that too, but I think we need to keep this consistent with other functions in this interface, e.g. MatchAgainstDeploymentAndNetworkFlow.

File: sensor/common/detector/queue/file_accesses.go (hunk 8 above)
Comment (Stringy) [reply sensor/common/detector/queue/file_accesses.go:10]: I'm sure I replied to this earlier, but it doesn't appear to have saved (apologies if this is somehow duplicated)

I'm not sure I fully understand what you mean or which storage proto we're using for any internal API. Do you mean the sending of storage.Alert to Central?

I think from a compatibility perspective, we're not really changing the API between Sensor and Central, since we're still just sending `storage.Alert`s, and the additional fields should be safely covered by gRPC compatibility guarantees. Though having said that I do wonder if there's a problem with sending the policies with the new critera to Sensor (e.g. new Central with the critera, and old Sensor without.) I'll try to investigate, but that will fall out of the scope of this PR.

We've also discussed Capabilities and that might help a lot with the compatibility (I've added https://issues.redhat.com/ROX-31892)

File: pkg/detection/runtime/utils.go (hunk 4 above)
Comment (clickboo) [reply pkg/detection/runtime/utils.go:35]: Discussed on slack: Condition should be || (and I'd flip the order)

File: pkg/features/list.go [hunk 11]
```diff
@@ -108,7 +108,7 @@ var (
 	KnownExploitedVulnerabilities = registerFeature("Display CISA Known Exploited Vulnerabilities (KEV) data", "ROX_CISA_KEV")
 
 	// SensitiveFileActivity enables monitoring of sensitive files.
-	SensitiveFileActivity = registerFeature("Enable sensitive file monitoring", "ROX_SENSITIVE_FILE_ACTIVITY")
+	SensitiveFileActivity = registerFeature("Enable sensitive file monitoring", "ROX_SENSITIVE_FILE_ACTIVITY", enabled)
```
Comment (JoukoVirtanen) [inline pkg/features/list.go:111]: Do you intend to merge this with the feature flag enabled by default?

Comment (clickboo) [reply pkg/features/list.go:111]: Based on his commit [9c218e1](https://github.com/stackrox/stackrox/commit/9c218e109856695b3802787639cfa9b97dace33c) @Stringy seems to have enabled it temporarily, most likely while debugging the pipeline issue.

**Comment by JoukoVirtanen** [comment]
I think you need to call the `Pause` and `Resume` methods for your new queue and call them in `Notify`

File: central/detection/alertmanager/alert_manager_impl.go [hunk 12]
```diff
@@ -229,12 +237,55 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+func mergeFileAccessViolations(oldAlert, newAlert *storage.Alert) bool {
+	if len(newAlert.GetFileAccessViolation().GetAccesses()) == 0 {
+		return false
+	}
+
+	if len(oldAlert.GetFileAccessViolation().GetAccesses()) >= maxRunTimeViolationsPerAlert {
+		return false
+	}
+
+	newFileAccesses := oldAlert.GetFileAccessViolation().GetAccesses()
```
Comment (clickboo) [inline central/detection/alertmanager/alert_manager_impl.go]: oldFileAccesses?

File: central/detection/alertmanager/alert_manager_impl.go [hunk 13]
```diff
@@ -229,12 +237,55 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
 	return mergeAlertsByLatestFirst(old, new, storage.Alert_Violation_NETWORK_FLOW)
 }
 
+func mergeFileAccessViolations(oldAlert, newAlert *storage.Alert) bool {
+	if len(newAlert.GetFileAccessViolation().GetAccesses()) == 0 {
```
Comment (clickboo) [inline central/detection/alertmanager/alert_manager_impl.go]: fileAccesses = newAlert.GetFileAccessViolation().GetAccesses()

numAccesses = len(fileAccesses)

if numAccesses == 0 || numAccesses >= maxRunTimeViolationsPerAlert { 
return false
}

File: central/detection/alertmanager/alert_manager_impl.go [hunk 14]
```diff
@@ -229,12 +237,55 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+		return false
+	}
+
+	newFileAccesses := oldAlert.GetFileAccessViolation().GetAccesses()
+	lastAccessTime, err := lastFileTime(oldAlert.GetFileAccessViolation().GetAccesses())
+	if err != nil {
+		log.Errorf(
+			"Failed to merge alerts. "+
+				"New alert %s (policy=%s) has %d file accesses and old alert %s (policy=%s) has %d file accesses: %v",
+			newAlert.GetId(), newAlert.GetPolicy().GetName(), len(newAlert.GetProcessViolation().GetProcesses()),
```
Comment (clickboo) [inline central/detection/alertmanager/alert_manager_impl.go]: numAccesses

File: central/detection/alertmanager/alert_manager_impl.go [hunk 15]
```diff
@@ -229,12 +237,55 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+			oldAlert.GetId(), oldAlert.GetPolicy().GetName(), len(oldAlert.GetFileAccessViolation().GetAccesses()), err,
+		)
+		// At this point, we know that the new alert has non-zero process violations but it cannot be merged.
```
Comment (clickboo) [inline central/detection/alertmanager/alert_manager_impl.go]: nit: zero file access violations (copy pasta!), more instances in comments below

File: sensor/common/filesystem/utils/utils.go [hunk 16]
```diff
@@ -3,7 +3,7 @@ package utils
 import "github.com/stackrox/rox/generated/storage"
 
 func IsNodeFileAccess(fileAccess *storage.FileAccess) bool {
-	return fileAccess.GetProcess().GetDeploymentId() != ""
+	return fileAccess.GetProcess().GetDeploymentId() == ""
```
Comment (clickboo) [inline sensor/common/filesystem/utils/utils.go:6]: Good catch

File: pkg/features/list.go (hunk 11 above)
Comment (Stringy) [reply pkg/features/list.go:111]: yeah, I'll revert this before merging

File: central/detection/alertmanager/alert_manager_impl.go [hunk 17]
```diff
@@ -229,12 +237,59 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+	mergedFileAccesses := oldViolations.GetAccesses()
+	lastAccessTime, err := lastFileTime(mergedFileAccesses)
+	if err != nil {
+		log.Errorf(
+			"Failed to merge alerts. "+
+				"New alert %s (policy=%s) has %d file accesses and old alert %s (policy=%s) has %d file accesses: %v",
+			newAlert.GetId(), newAlert.GetPolicy().GetName(), numNewAccesses,
+			oldAlert.GetId(), oldAlert.GetPolicy().GetName(), numOldAccesses, err,
+		)
+		// At this point, we know that the new alert has non-zero file violations but it cannot be merged.
```
Comment (JoukoVirtanen) [inline central/detection/alertmanager/alert_manager_impl.go:262]: Put a comment above the function declaration with an explanation as to what this function does and returns.

**Comment by JoukoVirtanen** [comment]
Is manual testing also a good idea?

File: central/detection/alertmanager/alert_manager_impl.go [hunk 18]
```diff
@@ -229,12 +237,59 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+	if !hasNewAccesses {
+		return hasNewAccesses
+	}
+
+	if len(mergedFileAccesses) > maxRunTimeViolationsPerAlert {
+		mergedFileAccesses = mergedFileAccesses[:maxRunTimeViolationsPerAlert]
+	}
+
+	newAlert.FileAccessViolation.Accesses = mergedFileAccesses
+	printer.UpdateFileAccessAlertViolationMessage(newAlert.GetFileAccessViolation())
```
Comment (JoukoVirtanen) [inline central/detection/alertmanager/alert_manager_impl.go:284]: `UpdateFileAccessAlertViolationMessage` was created in another PR, so I think it is outside of the scope of this PR, but I have a comment on it. If the cardinality is greater than 10, only the number of sensitive files accessed is printed. Wouldn't it be better if the first ten files were printed and then also printed that there were x number of files that were not printed?

Comment (Stringy) [reply central/detection/alertmanager/alert_manager_impl.go:284]: It's an optimisation for displaying a summary of the violations, and we keep all the file accesses so we're not losing any information. I think it's probably fine as is (and I suspect for 10 files is too many anyway, but I kept it at 10 to be consistent with the process violation messages)

File: central/detection/alertmanager/alert_manager_impl.go [hunk 19]
```diff
@@ -229,12 +237,59 @@ func mergeNetworkFlowViolations(old, new *storage.Alert) bool {
...
+			mergedFileAccesses = append(mergedFileAccesses, access)
+		}
+	}
+	// If there are no new accesses, we'll just use the old alert.
+	if !hasNewAccesses {
+		return hasNewAccesses
+	}
+
+	if len(mergedFileAccesses) > maxRunTimeViolationsPerAlert {
+		mergedFileAccesses = mergedFileAccesses[:maxRunTimeViolationsPerAlert]
```
Comment (Stringy) [inline central/detection/alertmanager/alert_manager_impl.go]: I think this is consistent in the sense that we prioritise old accesses (because we start with the list of old accesses, and then append new events.) But I do agree that keeping the new ones is probably more valuable. I can also fix the process violations too. @clickboo you might have more context on the historical reasons for this, WDYT?

File: central/detection/lifecycle/manager_impl.go [hunk 20]
```diff
@@ -487,6 +487,19 @@ func (m *managerImpl) HandleResourceAlerts(clusterID string, alerts []*storage.A
 	return nil
 }
 
+func (m *managerImpl) HandleNodeAlerts(clusterID string, alerts []*storage.Alert, stage storage.LifecycleStage) error {
```
Comment (JoukoVirtanen) [inline central/detection/lifecycle/manager_impl.go:490]: Maybe just pass in alertResults.

**Comment by JoukoVirtanen** [comment]
> I think this is consistent in the sense that we prioritise old accesses (because we start with the list of old accesses, and then append new events.) But I do agree that keeping the new ones is probably more valuable. I can also fix the process violations too. @clickboo you might have more context on the historical reasons for this, WDYT?

I think this is a response to one of my comments, that I deleted. I thought this was inconsistent, but I it is consistent. I probably misread `mergedFileAccesses = mergedFileAccesses[:maxRunTimeViolationsPerAlert]` as `mergedFileAccesses = mergedFileAccesses[maxRunTimeViolationsPerAlert:]`, or something. Having said that, I think it makes more sense to prioritize new alerts.

**Comment by JoukoVirtanen** [comment]
Feel free to resolve my comments and merge without my approval as I will be out of office.

File: sensor/common/filesystem/pipeline/pipeline.go [hunk 21]
```diff
@@ -53,6 +51,10 @@ func (p *Pipeline) translate(fs *sensorAPI.FileActivity) *storage.FileAccess {
 		Timestamp: fs.GetTimestamp(),
 	}
 
+	if fs.GetProcess().GetContainerId() == "" {
+		access.Hostname = fs.GetHostname()
+	}
```
Comment (lvalerom) [inline sensor/common/filesystem/pipeline/pipeline.go]: [nit] isn't this redundant? I think `Hostname` is already set in the previous lines.

File: pkg/env/sensor.go [hunk 22]
```diff
@@ -78,6 +78,13 @@ var (
 	// Multiple items can hold a pointer to the same object (e.g. same Deployment) so these numbers are pessimistic because we assume all items hold different objects.
 	DetectorDeploymentBufferSize = RegisterIntegerSetting("ROX_SENSOR_DETECTOR_DEPLOYMENT_BUFFER_SIZE", 20000)
 
+	// DetectorFileAccessBufferSize size indicates how many file access will be kept in Sensor while offline in the detector.
+	// 1 Item in the buffer = ~1000 bytes
```
Comment (lvalerom) [inline pkg/env/sensor.go:82]: [nit] This is probably fine but did you measure this? I remember doing some rough estimations on the other Item sizes.

File: pkg/detection/runtime/utils.go [hunk 23]
```diff
@@ -25,6 +25,37 @@ func constructProcessAlert(policy *storage.Policy, deployment *storage.Deploymen
 	return alert
 }
 
+// constructFileAccessAlert constructs an alert for a FileAccess violation.
+// It assumes one of node or deployment is non-nil and constructs an alert based
+// on this.
+func constructFileAccessAlert(policy *storage.Policy, node *storage.Node, deployment *storage.Deployment, violations booleanpolicy.Violations) *storage.Alert {
+	if len(violations.AlertViolations) == 0 && violations.FileAccessViolation == nil {
```
Comment (lvalerom) [inline pkg/detection/runtime/utils.go]: Does it make sense to have `len(violations.AlertViolations) = 0` and `violatiosn.FileAccessViolations != nil` and still create alerts (or the other way around)? If not, then I'd change this to an `or`.

**Review by lvalerom — APPROVED** [review]
LGTM

File: sensor/common/filesystem/pipeline/pipeline.go (hunk 21 above)
Comment (Stringy) [reply sensor/common/filesystem/pipeline/pipeline.go]: yeah not sure why this is here, I'll remove it

File: pkg/detection/runtime/utils.go (hunk 23 above)
Comment (Stringy) [reply pkg/detection/runtime/utils.go]: given the way this is currently used, it should only be `len(violations.AlertViolations) == 0 && violations.FileAccessViolations != nil` so I think I'll just simplify this down to `violations.FileAccessViolations == nil` to avoid confusion.

File: pkg/env/sensor.go (hunk 22 above)
Comment (Stringy) [reply pkg/env/sensor.go:82]: I haven't measured, but assumed this would be similar because the structure itself is similar to the others (it contains a context.Context + four pointers, so ~40 bytes actual size as pointed out below)

Comment (lvalerom) [reply pkg/env/sensor.go:82]: My logic behind those numbers was: The item itself is roughly 40 bytes but the structs are cloned and allocated somewhere. So even though the item is roughly 40 bytes, it represents a memory consumption of ~1000 bytes in the cases of deployments with netflows and PIs. I don't know how much memory the `FileAccess` + `Deployment` or `FileAccess` + `Node` use. If you think it should be similar then that's ok. I was just curious 🙂
